<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <arg name="world" default="racetrack_decorated_2"/>
    <arg name="paused" default="false"/>
    <arg name="use_sim_time" default="true"/>
    <arg name="gui" default="true"/>
    <arg name="debug" default="false"/>
    <arg name="verbose" default="true"/>
    <arg name="use_gpu" default="true"/>
    <arg name="joystick_type" default="xbox360"/>
    <arg name="visualize_lidar" default="false"/>
    <arg name="emergency_stop" default="true"/>

    <!-- Mode Override
        1 forces manual driving
        2 forces autonomous driving
        0 uses user input to enable the modes
    -->
    <arg name="mode_override" default="0"/>

    <param name="world_name" value="$(arg world)"/>

    <include file="$(find racer_world)/launch/racer_gazebo.launch">
        <arg name="world" value="$(arg world)"/>
        <arg name="paused" value="$(arg paused)"/>
        <arg name="use_sim_time" value="$(arg use_sim_time)"/>
        <arg name="gui" value="$(arg gui)"/>
        <arg name="debug" value="$(arg debug)"/>
        <arg name="verbose" value="$(arg verbose)"/>
        <arg name="use_gpu" value="$(arg use_gpu)"/>
        <arg name="visualize_lidar" value="$(arg visualize_lidar)"/>
    </include>
    
    <!-- transform laserscan to cartesian -->
    <include file="$(find car_tf)/launch/car_transformer.launch" />

    <!-- zed camera launch file -->
    <!--<include file="./launch/zed.launch"/>-->

    <!-- laserscan from kinect -->
    <node pkg="depthimage_to_laserscan" type="depthimage_to_laserscan" name="depthimage_to_laserscan" respawn="true">
        <param name="scan_height" value="5"/> 
        <param name="scan_time" value="0.0333"/>  
        <param name="range_min" value="0.3"/>  
        <param name="range_max" value="20"/> 
        <remap from="scan" to="scan_camera"/>
        <param name="output_frame_id" value="laser"/>
        <remap from="image" to="/racer/camera1/depth/image_raw"/>
    </node> <!-- we convert this laserscan back to a pointcloud for easier merging -->
    <include file="$(find car_tf)/launch/car_transformer.launch">
        <arg name="topic_input" value="/scan_camera" />
        <arg name="topic_output" value="/scan/cam/cartesian" />
    </include>
    <!-- why the hassle you may ask...
         the depthimage_to_laserscan package offers a lot of options which are really useful for our case.
         if it turns out that this conversion is to computationally intensive, we have to find another way
         e.g. directly use the cameras pointcloud. we may have to do that in the future anyways...
        -->
    <include file="$(find cam_lidar_fusion)/launch/cam_lidar_fusion.launch">
        <arg name="topic_input_cam" value="/scan/cam/cartesian"/>
        <arg name="topic_input_lidar" value="/scan/lidar/cartesian"/>
        <arg name="topic_output_fused" value="/scan/fusion/cartesian"/>
    </include>
    
    <include file="$(find boxing)/launch/boxing.launch">
        <arg name="topic_input_cloud" value="/scan/fusion/cartesian" />
        <arg name="topic_voxels" value="/scan/voxels" />
    </include>

    <include file="$(find voxel_classifier)/launch/voxel_classifier.launch">
        <arg name="topic_input_voxels" value="/scan/voxels" />
        <arg name="topic_output_clusters" value="/scan/clusters" />
    </include>

    <include file="$(find wall_detection)/launch/wall_detection.launch">
        <arg name="topic_input_clusters" value="/scan/clusters" />
        <arg name="topic_output_walls" value="/obstacles/walls" />
        <arg name="topic_output_obstacles" value="/obstacles/obstacles" />
    </include>

    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find racer_world)/launch/rviz_config.rviz"/>

    <node name="lap_timer" pkg="simulation_tools" type="lap_timer.py" output="screen"/>
    <node pkg="gazebo_telemetry" type="gazebo_state_telemetry.py" name="gazebo_state_telemetry" output="screen"/>
    <include file="$(find wallfollowing5)/launch/autonomous_driving.launch"/>

    <group if="$(arg emergency_stop)">
        <include file="$(find emergency_stop)/launch/emergency_stop.launch"/>
    </group>

    <include file="$(find vesc_sim)/launch/vesc_sim.launch"/>
    <include file="$(find car_control)/launch/car_control.launch">
        <arg name="mode_override" value="$(arg mode_override)"/>
    </include>
    <include file="$(find teleoperation)/launch/remote_control.launch">
        <arg name="joystick_type" value="$(arg joystick_type)"/>
    </include>

    <include file="$(find imu_velocity)/launch/car_imu_velocity.launch" />

    <node pkg="simulation_tools" type="crash_detector" name="crash_detector" output="screen">
    </node>
    <node pkg="tf" type="static_transform_publisher" name="depth_rotation" args="0.18 0 0.14 -1.570796327 0 -1.570796327 /base_link /camera 10" />

    <node pkg="simulation_tools" type="log_stats.py" name="log_stats" output="screen" args="sim-current"/>
</launch>
